modules/ismp/clients/tendermint/verifier/src/types.rs

```diff
@@ -0,0 +1,268 @@
+use std::time::Duration;
+
+use crate::{SignedHeader, Validator};
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct TrustedState {
+	/// Chain ID
+	pub chain_id: String,
+	/// Block height
+	pub height: u64,
+	/// Block timestamp
+	pub timestamp: u64,
+	/// Current validator set
+	pub validators: Vec<Validator>,
+	/// Next validator set
+	pub next_validators: Vec<Validator>,
+	/// Hash of the next validator set
+	pub next_validators_hash: [u8; 32],
+	/// Trusting period in seconds
+	pub trusting_period: u64,
+	/// Frozen height (if any)
+	pub frozen_height: Option<u64>,
+}
+
+impl TrustedState {
+	pub fn new(
+		chain_id: String,
+		height: u64,
+		timestamp: u64,
+		validators: Vec<Validator>,
+		next_validators: Vec<Validator>,
+		next_validators_hash: [u8; 32],
+		trusting_period: u64,
+	) -> Self {
+		Self {
+			chain_id,
+			height,
+			timestamp,
+			validators,
+			next_validators,
+			next_validators_hash,
+			trusting_period,
+			frozen_height: None,
+		}
+	}
+
+	pub fn is_frozen(&self) -> bool {
+		self.frozen_height.is_some()
+	}
+
+	pub fn trusting_period_duration(&self) -> Duration {
+		Duration::from_secs(self.trusting_period)
+	}
+
+	/// Validate the trusted state
+	pub fn validate(&self) -> Result<(), String> {
+		if self.chain_id.is_empty() {
+			return Err("Chain ID cannot be empty".to_string());
+		}
+		if self.height == 0 {
+			return Err("Height cannot be zero".to_string());
+		}
+		if self.timestamp == 0 {
+			return Err("Timestamp cannot be zero".to_string());
+		}
+		if self.validators.is_empty() {
+			return Err("Validator set cannot be empty".to_string());
+		}
+		if self.next_validators.is_empty() {
+			return Err("Next validator set cannot be empty".to_string());
+		}
+		if self.trusting_period == 0 {
+			return Err("Trusting period cannot be zero".to_string());
+		}
+		Ok(())
+	}
+
+	/// Create a frozen trusted state
+	pub fn freeze_at_height(mut self, height: u64) -> Self {
+		self.frozen_height = Some(height);
+		self
+	}
+
+	/// Check if the state is valid for a given height
+	pub fn is_valid_for_height(&self, height: u64) -> bool {
+		match self.frozen_height {
+			Some(frozen_height) => height < frozen_height,
+			None => true,
+		}
+	}
+}
+
+impl Default for TrustedState {
+	fn default() -> Self {
+		Self {
+			chain_id: "test-chain".to_string(),
+			height: 1,
+			timestamp: 0,
+			validators: Vec::new(),
+			next_validators: Vec::new(),
+			next_validators_hash: [0u8; 32],
+			trusting_period: 3600, // 1 hour default
+			frozen_height: None,
+		}
+	}
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct ConsensusProof {
+	/// Signed header containing the block header and commit
+	pub signed_header: SignedHeader,
```

Comment:
The signed header needs to come with an ancestry(a list of signed headers from the trusted block height to the latest signed header). Tendermint has single-slot finality, but we would not be submitting proofs for each slot, hence the need for an ancestry check